presto-hive/src/main/java/com/facebook/presto/hive/HivePageSource.java

```diff
@@ -595,11 +621,98 @@ public void load(LazyBlock lazyBlock)
             if (block instanceof LazyBlock) {
                 block = ((LazyBlock) block).getBlock();
             }
-            Block coercedBlock = coercer.apply(block);
-            lazyBlock.setBlock(coercedBlock);
+            lazyBlock.setBlock(coercer.apply(block));
+
+            // clear reference to loader to free resources, since load was successful
+            block = null;
+        }
+    }
+
+    private static final class RowFilterLazyBlockLoader
+            implements LazyBlockLoader<LazyBlock>
+    {
+        private Block block;
+        private final IntArrayList rowsToKeep;
+
+        public RowFilterLazyBlockLoader(Block block, IntArrayList rowsToKeep)
+        {
+            this.block = requireNonNull(block, "block is null");
+            this.rowsToKeep = requireNonNull(rowsToKeep, "rowsToKeep is null");
+        }
+
+        @Override
+        public void load(LazyBlock lazyBlock)
+        {
+            if (block == null) {
+                return;
+            }
+
+            if (block instanceof LazyBlock) {
+                block = ((LazyBlock) block).getBlock();
+            }
+            lazyBlock.setBlock(filterBlock(block, rowsToKeep));
 
             // clear reference to loader to free resources, since load was successful
             block = null;
         }
+
+        private static Block filterBlock(Block block, IntArrayList rowsToKeep)
+        {
+            if (block instanceof RunLengthEncodedBlock) {
```

Comment:
This is what `getPosition` should already be doing in these implementations.

Reply:
The implementation of `getPositions` in `RunLengthEncodedBlock` should already be doing this.  Also, every's getPositions, should already be implemented using masking.